src/libexpr/nixexpr.cc

```diff
@@ -522,6 +583,51 @@ std::string ExprLambda::showNamePos(const EvalState & state) const
 
 
 
+/* Position table. */
+
+Pos PosTable::operator[](PosIdx p) const
+{
+    if (p.id == 0)
+        return {};
+
+    const auto idx = p.id - 1;
+    /* we want the last key <= idx, so we'll take prev(first key > idx).
+        this is guaranteed to never rewind origin.begin because the first
+        key is always 0. */
+    const auto pastOrigin = origins.upper_bound(idx);
+    const auto & origin = std::prev(pastOrigin)->second;
+    const auto offset = idx - origin.offset;
+
+    Pos result{0, 0, origin.origin};
+    auto lines = this->lines.lock();
+    auto linesForInput = (*lines)[origin.offset];
+
+    if (linesForInput.empty()) {
+        auto source = result.getSource().value_or("");
+        const char * begin = source.data();
+        for (Pos::LinesIterator it(source), end; it != end; it++)
+            linesForInput.push_back(it->data() - begin);
+    }
+    // as above: the first line starts at byte 0 and is always present
+    auto lineStartOffset = std::prev(
+        std::upper_bound(linesForInput.begin(), linesForInput.end(), offset));
+
+    result.line = 1 + (lineStartOffset - linesForInput.begin());
+    result.column = 1 + (offset - *lineStartOffset);
+    return result;
+}
+
+Pos::Origin PosTable::originOf(PosIdx p) const
+{
+    if (p.id == 0)
+        return std::monostate{};
+
+    // see above for justification.
```

Comment:
```suggestion
    // see `PosTable::operator[]` for justification.
```

New methods could be added, making "above" ambiguous.